### PR TITLE
[SPARK-17212][SQL] TypeCoercion supports widening conversion between DateType and TimestampType

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Currently, type-widening does not work between `TimestampType` and `DateType`.

This applies to `SetOperation`, `Union`, `In`, `CaseWhen`, `Greatest`,  `Leatest`, `CreateArray`, `CreateMap`, `Coalesce`, `NullIf`, `IfNull`, `Nvl` and `Nvl2`, .

This PR adds the support for widening `DateType` to `TimestampType` for them.

For a simple example, 

**Before**

``` scala
Seq(Tuple2(new Timestamp(0), new Date(0))).toDF("a", "b").selectExpr("greatest(a, b)").show()
```

shows below:

```
cannot resolve 'greatest(`a`, `b`)' due to data type mismatch: The expressions should all have the same type, got GREATEST(timestamp, date)
```

or union as below:

``` scala
val a = Seq(Tuple1(new Timestamp(0))).toDF()
val b = Seq(Tuple1(new Date(0))).toDF()
a.union(b).show()
```

shows below:

```
Union can only be performed on tables with the compatible column types. DateType <> TimestampType at the first column of the second table;
```

**After**

``` scala
Seq(Tuple2(new Timestamp(0), new Date(0))).toDF("a", "b").selectExpr("greatest(a, b)").show()
```

shows below:

```
+----------------------------------------------------+
|greatest(CAST(a AS TIMESTAMP), CAST(b AS TIMESTAMP))|
+----------------------------------------------------+
|                                1969-12-31 16:00:...|
+----------------------------------------------------+
```

or union as below:

``` scala
val a = Seq(Tuple1(new Timestamp(0))).toDF()
val b = Seq(Tuple1(new Date(0))).toDF()
a.union(b).show()
```

shows below:

```
+--------------------+
|                  _1|
+--------------------+
|1969-12-31 16:00:...|
|1969-12-31 00:00:...|
+--------------------+
```
## How was this patch tested?

Unit tests in `TypeCoercionSuite`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Cc @cloud-fan and @petermaxlee  could you please take a look for this one?

**Comment by cloud-fan** [comment]
Have you checked with other databases? It looks reasonable to do this but I'd like to know how other databases deal with this case.

**Comment by HyukjinKwon** [comment]
I have only checked Hive it seems it does not handle this case. I will check others and be back soon.

**Comment by HyukjinKwon** [comment]
MySQL and PostgreSQL support this

**MySQL**

- Greatest/leastest

```sql
mysql> SELECT GREATEST(CAST("1990-02-24 12:00:00" AS DATETIME), CAST("1990-02-25" AS DATE));
+-------------------------------------------------------------------------------+
| GREATEST(CAST("1990-02-24 12:00:00" AS DATETIME), CAST("1990-02-25" AS DATE)) |
+-------------------------------------------------------------------------------+
| 1990-02-25 00:00:00                                                           |
+-------------------------------------------------------------------------------+
```

- Union

```sql
mysql> SELECT CAST("1990-02-24 12:00:00" AS DATETIME) UNION SELECT CAST("1990-02-24" AS DATE);
+-----------------------------------------+
| CAST("1990-02-24 12:00:00" AS DATETIME) |
+-----------------------------------------+
| 1990-02-24 12:00:00                     |
| 1990-02-24 00:00:00                     |
+-----------------------------------------+
```

**PostgreSQL**

- Greatest/leatest

```sql
postgres=# SELECT GREATEST(CAST('1990-02-24 12:00:00' AS TIMESTAMP), CAST('1990-02-25' AS DATE));
      greatest
---------------------
 1990-02-25 00:00:00
(1 row)
```

- Union

```sql
postgres=# SELECT CAST('1990-02-24 12:00:00' AS TIMESTAMP) UNION SELECT CAST('1990-02-24' AS DATE);
      timestamp
---------------------
 1990-02-24 00:00:00
 1990-02-24 12:00:00
(2 rows)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -134,6 +134,8 @@ object TypeCoercion {
       Some(DecimalPrecision.widerDecimalType(DecimalType.forType(t), d))
     case (_: FractionalType, _: DecimalType) | (_: DecimalType, _: FractionalType) =>
       Some(DoubleType)
+    case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: shall we put it in `findTightestCommonTypeOfTwo`?

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Ah, sure!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 2]
```diff
@@ -188,6 +188,7 @@ class TypeCoercionSuite extends PlanTest {
     // TimestampType
     widenTest(NullType, TimestampType, Some(TimestampType))
     widenTest(TimestampType, TimestampType, Some(TimestampType))
+    widenTest(DateType, TimestampType, Some(TimestampType))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: I think we only need this test

Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: Yeap, fixed.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**